README.md
[![Build Status](https://travis-ci.org/Factoshi/graff.svg?branch=master)](https://travis-ci.org/Factoshi/graff)

# Graff

Graff is a GraphQL wrapper for the factomd RPC API - it allows a client to query factomd using GraphQL.

Note: this is an early beta. Fields or types may be deprecated or removed at short notice.

## Requirements

-   Either Docker or Node.js 12 or higher.
-   A running factomd instance.

## Installation

### Docker (recommended)

Simple installation (can be configured with environment variables, see below). Replace `FACTOMD_HOST` environment variable with the URI of your factomd instance.

```
docker run -d --name graff -p 8032:8032 -e FACTOMD_HOST=factomd factoshi/graff:latest
```

And that's it!

If you want some docker-compose recipe ideas, check out the [recipes directory](recipes). Pull requests for more are appeciated!

### Node

Clone this repo and cd into the project.

```
git clone https://github.com/Factoshi/graff.git
cd graff
```

Install and compile.

```
npm install --production
npm run build
```

Run it using default configuration (see below for config options).

```
NODE_ENV=production node build/index.js
```

## Configuration

Graff comes with extensive configuration options that are all delivered using environment variables. Note: defaults are for production environments (i.e. where NODE_ENV=production), and may vary for non-production environments.

### Factomd

| Variable         | Description                                       | Default   |
| ---------------- | ------------------------------------------------- | --------- |
| FACTOMD_HOST     | Host of the factomd instance.                     | localhost |
| FACTOMD_PORT     | Port of the factomd API.                          | 8088      |
| FACTOMD_PATH     | Path to the factomd v2 API.                       | /v2       |
| FACTOMD_PROTOCOL | Protocol to connect to API. Can be http or https. | http      |
| FACTOMD_USER     | User for RPC authentication.                      |           |
| FACTOMD_PASSWORD | Password for RPC authentication.                  |           |

### Redis

Graff uses an in-memory LRU cache to speed up requests and reduce load on factomd. If you are using multiple Graff instances, they are able to share the same Redis cache to save memory.

| Variable       | Description                        | Default   |
| -------------- | ---------------------------------- | --------- |
| REDIS_HOST     | Host of the redis instance.        | localhost |
| REDIS_PORT     | Port of the redis instance.        | 6379      |
| REDIS_PASSWORD | Password for Redis authentication. |           |
| REDIS_FAMILY   | IP version to use. Can be 4 or 6.  | 4         |
| REDIS_DB       | Database to connect to.            | 0         |

### GraphQL Server

| Variable       | Description                          | Default |
| -------------- | ------------------------------------ | ------- |
| GQL_PLAYGROUND | Boolean to enable GQL playground.    | false   |
| GQL_INTROSPEC  | Boolean to enable introspection.     | false   |
| GQL_PORT       | Port where the server should listen. | 8032    |
| KEEP_ALIVE     | Frequency of websocket pings (ms).   | 30000   |

### DoS protection

GraphQL is vulnerable to DoS attacks via malicious queries. These variables let you configure the DoS protection settings.

| Variable        | Description                                                                                                                 | Default |
| --------------- | --------------------------------------------------------------------------------------------------------------------------- | ------- |
| MAX_PAGE_LENGTH | Maximum page length for paginated types.                                                                                    | 150     |
| MAX_QUERY_DEPTH | Maximum query depth (i.e. limit of nested queries).                                                                         | 7       |
| MAX_COMPLEXITY  | Maximum query complexity. See [this repo](https://github.com/4Catalyzer/graphql-validation-complexity) for more information | 1500    |


changelog.md
# Changelog

## [0.2.1]

### Upgrade

-   Fix vulnerable dependencies.
-   Fix query receipt test to remove bitcoin assertions.

## [0.2.0]

### Changed

-   Remove artifical millisecond timestamps for directory block and entry block.

## [0.1.2]

### Removed

-   Websocket keep alive option.

## [0.1.2]

### Added

-   Changelog.
-   Default websocket keep alive of 30s and environment variable to set own.


src/contants.ts
/**
 * Environment variables.
 */
const env = process.env;
// factomd
export const FACTOMD_HOST = env.FACTOMD_HOST || 'localhost';
export const FACTOMD_PORT = env.FACTOMD_PORT ? parseInt(env.FACTOMD_PORT) : 8088;
export const FACTOMD_PATH = env.FACTOMD_PATH || '/v2';
export const FACTOMD_USER = env.FACTOMD_USER;
export const FACTOMD_PASSWD = env.FACTOMD_PASSWD;
export const FACTOMD_PROTOCOL = env.FACTOMD_PROTOCOL || 'http';

// Redis config
export const REDIS_PORT = env.REDIS_PORT ? parseInt(env.REDIS_PORT) : 6379;
export const REDIS_HOST = env.REDIS_HOST;
export const REDIS_PASSWD = env.REDIS_PASSWD;
export const REDIS_FAMILY = env.REDIS_FAMILY === '6' ? 6 : 4; // IP version
export const REDIS_DB = env.REDIS_DB ? parseInt(env.REDIS_DB) : 0;

// Server config
export const GQL_PLAYGROUND = env.GQL_PLAYGROUND
    ? env.GQL_PLAYGROUND === 'true'
    : env.NODE_ENV !== 'production';
export const GQL_PORT = env.GQL_PORT
    ? parseInt(env.GQL_PORT)
    : env.NODE_ENV !== 'production'
    ? 4000
    : 8032;
export const GQL_INTROSPEC = env.GQL_INTROSPEC
    ? env.GQL_INTROSPEC === 'true'
    : env.NODE_ENV !== 'production';
export const MAX_PAGE_LENGTH = env.MAX_PAGE_LENGTH ? parseInt(env.MAX_PAGE_LENGTH) : 150;
export const MAX_QUERY_DEPTH = env.MAX_QUERY_DEPTH ? parseInt(env.MAX_QUERY_DEPTH) : 7;
export const MAX_COMPLEXITY = env.MAX_COMPLEXITY ? parseInt(env.MAX_COMPLEXITY) : 1500;

// Subscription channels
export enum Channel {
    NewAdminBlock = 'NEW_ADMIN_BLOCK',
    NewDirectoryBlock = 'NEW_DIRECTORY_BLOCK',
    NewChains = 'NEW_CHAINS',
    NewEntryCreditBlock = 'NEW_ENTRY_CREDIT_BLOCK',
    NewFactoidBlock = 'NEW_FACTOID_BLOCK',
    newEntryBlock = 'NEW_ENTRY_BLOCK'
}

// Mutation methods
export enum MutationMethod {
    CommitChain = 'commit-chain',
    RevealChain = 'reveal-chain',
    CommitEntry = 'commit-entry',
    RevealEntry = 'reveal-entry',
    SubmitTx = 'factoid-submit'
}
